content/D_AlgebraicUnderpinnings.md

```diff
@@ -0,0 +1,600 @@
+    {
+        index++;
+        return c;
+    });
+
+result.Subscribe(
+    c => Console.WriteLine("Received {0} at index {1}", c, index),
+    () => Console.WriteLine("completed"));
+
+result.Subscribe(
+    c => Console.WriteLine("Also received {0} at index {1}", c, index),
+    () => Console.WriteLine("2nd completed"));
+```
+
+Output
+
+```
+Received A at index 0
+Received B at index 1
+Received C at index 2
+completed
+Also received A at index 3
+Also received B at index 4
+Also received C at index 5
+2nd completed
+```
+
+Now the second person's output is clearly nonsense. They will be expecting index values to be 0, 1 and 2 but get 3, 4 and 5 instead. I have seen far more sinister versions of side effects in code bases. The nasty ones often modify state that is a Boolean value e.g. `hasValues`, `isStreaming` etc.
+
+In addition to creating potentially unpredictable results in existing software, programs that exhibit side effects are far more difficult to test and maintain. Future refactoring, enhancements or other maintenance on programs that exhibits side effects are far more likely to be brittle. This is especially so in asynchronous or concurrent software.
+
+## Composing data in a pipeline
+
+The preferred way of capturing state is as part of the information flowing through the pipeline of Rx operators making up your subscription. Ideally, we want each part of the pipeline to be independent and deterministic. That is, each function that makes up the pipeline should have its inputs and output as its only state. To correct our example we could enrich the data in the pipeline so that there is no shared state. This would be a great example where we could use the `Select` overload that exposes the index.
+
+```cs
+IObservable<int> source = Observable.Range(0, 3);
+IObservable<(int Index, char Letter)> result = source.Select(
+    (idx, value) => (Index: idx, Letter: (char) (value + 65)));
+
+result.Subscribe(
+    x => Console.WriteLine($"Received {x.Letter} at index {x.Index}"),
+    () => Console.WriteLine("completed"));
+
+result.Subscribe(
+    x => Console.WriteLine($"Also received {x.Letter} at index {x.Index}"),
+    () => Console.WriteLine("2nd completed"));
+```
+
+Output:
+
+```
+Received A at index 0
+Received B at index 1
+Received C at index 2
+completed
+Also received A at index 0
+Also received B at index 1
+Also received C at index 2
+2nd completed
+```
+
+Thinking outside of the box, we could also use other features like `Scan` to achieve similar results. Here is an example.
+
+```csharp
+var result = source.Scan(
+                new
+                {
+                    Index = -1,
+                    Letter = new char()
+                },
+                (acc, value) => new
+                {
+                    Index = acc.Index + 1,
+                    Letter = (char)(value + 65)
+                });
+```
+
+The key here is to isolate the state, and reduce or remove any side effects like mutating state.
```

Comment:
Could the appendix / book do with a summary / wrap up?

Reply:
I pondered this when I wrote it, but didn't write a summary for the following reasons:

1. none of the other appendices have a summary
2. appendices are by definition stuff that didn't really fit into a narrative structure, so it's quite common for them to stop abruptly; they're sort of like footnotes that got out of hand
3. this chapter exists mainly because Lee's original book had lots of little references to bits of maths that influenced Rx, but with no real explanation of what they meant, how they applied to Rx, or why you might care; this appendix grew out of that, so it really is just a bunch of thematically related facts that would have got in the way of the flow if mentioned inline; my attempts to come up with some sort of summary were all in no way an improvement on no summary at all.

As for a book wrap up... My assumption is that few people will read the entire thing cover to cover. Certainly the way I tend to read most technical books is that I'll read the first few chapters linearly, but that I usually reach a point where I start to become selective. So a wrap-up at the end feels like it's destined to go unread.